src/Http/Routing/src/Matching/DefaultEndpointSelector.cs

```diff
@@ -104,6 +119,104 @@ private static void ProcessFinalCandidates(
         }
     }
 
+    private static CandidateState? SelectMostSpecificEndpoint(List<CandidateState> candidates)
+    {
+        CandidateState? mostSpecific = null;
+        var highestSpecificity = -1;
+        var hasAmbiguity = false;
+
+        foreach (var candidate in candidates)
+        {
+            if (candidate.Endpoint is not RouteEndpoint routeEndpoint)
+            {
+                continue;
+            }
+
+            var specificity = CalculateConstraintSpecificity(routeEndpoint);
+
+            if (specificity > highestSpecificity)
+            {
+                highestSpecificity = specificity;
+                mostSpecific = candidate;
+                hasAmbiguity = false;
+            }
+            else if (specificity == highestSpecificity)
+            {
+                // Okay, note the ambiguity and continue trying
+                // to determine a higher level of specificity.
+                hasAmbiguity = true;
+            }
+        }
+
+        return hasAmbiguity ? null : mostSpecific;
+    }
+
+    private static int CalculateConstraintSpecificity(RouteEndpoint endpoint)
+    {
+        var specificity = 0;
+        var routePattern = endpoint.RoutePattern;
+
+        foreach (var parameter in routePattern.Parameters)
+        {
+            // We may have parameter without constraints, e.g. "id" in "/products/{id}"
+            if (parameter.ParameterPolicies?.Count > 0)
+            {
+                foreach (var policy in parameter.ParameterPolicies)
+                {
+                    if (policy.Content != null)
+                    {
+                        specificity += GetConstraintSpecificityWeight(policy.Content);
+                    }
+                }
+            }
+        }
+
+        return specificity;
+    }
+
+    private static int GetConstraintSpecificityWeight(string constraintName)
+    {
+        return constraintName.ToLowerInvariant() switch
+        {
+            // Strong typed constraints that are very restrictive and has
+            // the highest specificity
+            "guid" => 100,
```

Comment:
Using .Equals with a StringComparison to ignore casing might be more consistent than calling ToLowerInvariant for every constraint.